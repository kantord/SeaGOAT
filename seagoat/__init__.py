__version__ = "0.49.4"
