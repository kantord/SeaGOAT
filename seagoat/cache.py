import hashlib
import os
import pickle
from pathlib import Path
from typing import Generic
from typing import TypeVar

import appdirs

T = TypeVar("T")

CACHE_FORMAT_VERSION = 15


class Cache(Generic[T]):
    def __init__(self, cache_name: str, path: Path, initial_value: T):
        self._path = path
        self.data = initial_value
        self._cache_name = cache_name

    def load(self):
        try:
            with open(self._get_cache_file(), "rb") as cache_file:
                self.data = pickle.load(cache_file)
        except (FileNotFoundError, pickle.UnpicklingError, EOFError):
            pass

    def persist(self):
        with open(self._get_cache_file(), "wb") as cache_file:
            pickle.dump(self.data, cache_file)

    def _get_cache_file(self):
        return self.get_cache_folder() / self._cache_name

    def get_cache_folder(self):
        cache_folder = self._get_cache_root() / self._get_project_hash()
        cache_folder.mkdir(parents=True, exist_ok=True)

        return cache_folder

    def _get_cache_root(self):
        if "RUNNER_TEMP" in os.environ:
            return Path(os.environ["RUNNER_TEMP"])

        return Path(
            appdirs.user_cache_dir(
                "seagoat-pytest" if "PYTEST_CURRENT_TEST" in os.environ else "seagoat"
            )
        )

    def _get_project_hash(self):
        normalized_path = Path(self._path).expanduser().resolve()
        text = f"""
        Cache version: {CACHE_FORMAT_VERSION}
        Normalized path: {normalized_path}
        """

        return hashlib.sha256(text.encode()).hexdigest()
