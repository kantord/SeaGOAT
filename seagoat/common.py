IGNORED_BRANCHES = {"gh-pages"}
SUPPORTED_FILE_TYPES = {
    ".txt",
    ".md",
    ".py",
    ".c",
    ".cpp",
    ".h",
    ".hpp",
    ".ts",
    ".js",
    ".tsx",
    ".jsx",
    ".html",
}
