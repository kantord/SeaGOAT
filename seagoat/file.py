import hashlib
import time
from typing import List
from typing import Literal

from chromadb.errors import Dict


class File:
    def __init__(self, path: str, absolute_path: str):
        self.path = path
        self.absolute_path = absolute_path
        self._commit_times = set()
        self.commit_messages = set()

    def __repr__(self):
        return f"<File {self.path} {self.get_score()}>"

    def get_score(self):
        current_time = int(time.time())

        return sum(
            1000 / max(int((current_time - committed_date) / 86400) ** 2, 1)
            for committed_date in self._commit_times
        )

    def _add_commit_message(self, message: str):
        self.commit_messages.add(message)

    def add_commit(self, commit):
        self._commit_times.add(commit.committed_date)
        self._add_commit_message(commit.message)

    def get_metadata(self):
        commit_messages = "\n-".join(sorted(self.commit_messages))
        return f"""###
    Filename: {self.path}
    Commits:
{commit_messages}"""

    def _get_file_lines(self) -> Dict[int, str]:
        with open(self.absolute_path, "r", encoding="utf-8") as source_code_file:
            lines = {
                (i + 1): line
                for i, line in enumerate(source_code_file.read().splitlines())
            }

        return lines

    def _format_chunk_summary(self, relevant_lines: List[str]):
        truncated_lines = [line[:100] for line in relevant_lines]
        chunk = "\n".join(truncated_lines)
        chunk = chunk + self.get_metadata()

        return chunk

    def _get_context_lines(
        self, lines: Dict[int, str], line_number: int, direction: Literal[-1, 1]
    ) -> List[str]:
        context_lines = []
        for i in range(1, 6):
            current_line_number = line_number + (direction * i)
            current_line = lines.get(current_line_number)

            if current_line is None:
                break

            if direction == -1:
                context_lines = [current_line] + context_lines
            else:
                context_lines.append(current_line)

            if self._line_has_relevant_data(current_line):
                break

        return context_lines

    def _get_chunk_for_line(self, line_number: int, lines: Dict[int, str]):
        relevant_lines = (
            self._get_context_lines(lines, line_number, -1)
            + [lines[line_number]]
            + self._get_context_lines(lines, line_number, 1)
        )
        return FileChunk(self, line_number, self._format_chunk_summary(relevant_lines))

    def _line_has_relevant_data(self, line: str):
        return sum(c.isalnum() for c in line) > 3

    def get_chunks(self):
        try:
            lines = self._get_file_lines()
            return [
                self._get_chunk_for_line(line_number, lines)
                for line_number in lines.keys()
                if self._line_has_relevant_data(lines[line_number])
            ]

        except FileNotFoundError:
            return []


# pylint: disable=too-few-public-methods
class FileChunk:
    def __init__(self, parent: File, codeline: int, chunk: str):
        self.path = parent.path
        self.codeline = codeline
        self.chunk = chunk
        self.chunk_id = self._get_id()

    def _get_id(self):
        text = f"""
        Path: {self.path}
        Code line: {self.codeline}
        Chunk: {self.chunk}
        """
        return hashlib.sha256(text.encode()).hexdigest()
