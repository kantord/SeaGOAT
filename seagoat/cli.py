import os
from functools import cache

import click
import requests
from pygments import highlight
from pygments.formatters import TerminalFormatter
from pygments.lexers import get_lexer_for_filename
from pygments.lexers.javascript import JavascriptLexer
from pygments.lexers.javascript import TypeScriptLexer

from seagoat.server import get_server_info_file
from seagoat.server import load_server_info


def query_server(query, server_address):
    response = requests.get(f"{server_address}/query/{query}")
    try:
        response.raise_for_status()
    except requests.exceptions.HTTPError as err:
        print("HTTP error occurred:")
        print(f"Response code: {response.status_code}")
        print(f"Error: {err}")
        print(f"Response body: {response.text}")
        raise
    return response.json()["results"]


@cache
def get_highlighted_lines(file_name: str):
    with open(file_name, "r", encoding="utf-8") as source_code_file:
        code = source_code_file.read()

    if file_name.endswith(".md"):
        return code.splitlines()

    if file_name.endswith(".jsx"):
        lexer = JavascriptLexer()
    elif file_name.endswith(".tsx"):
        lexer = TypeScriptLexer()
    else:
        lexer = get_lexer_for_filename(file_name)

    result = highlight(code, lexer, TerminalFormatter())

    return result.splitlines()


def print_result_line(result, line, color_enabled):
    if color_enabled:
        highlighted_lines = get_highlighted_lines(str(result["full_path"]))
        print(
            f"{result['path']}:{click.style(str(line), bold=True)}:{highlighted_lines[line - 1]}"
        )
    else:
        for line_content in result["lines"]:
            if line_content["line"] == line:
                print(f"{result['path']}:{line}:{line_content['line_text']}")
                break


@click.command()
@click.argument("query")
@click.argument("repo_path", required=False, default=os.getcwd())
@click.option(
    "--no-color",
    is_flag=True,
    help="Disable formatting. Automatically enabled when part of a bash pipeline.",
)
def seagoat(query, repo_path, no_color):
    """
    Query your codebase for your QUERY in the Git repository REPO_PATH.

    Your query can either be text that you expect to find in a file,
    or a description of what you are looking for.

    When REPO_PATH is not specified, the current working directory is
    assumed to be the repository path.
    """
    _, __, ___, server_address = load_server_info(get_server_info_file(repo_path))
    results = query_server(query, server_address)

    color_enabled = os.isatty(0) and not no_color
    for result in results:
        for result_line in result.get("lines", []):
            print_result_line(result, result_line["line"], color_enabled)


if __name__ == "__main__":
    # pylint: disable-next=no-value-for-parameter
    seagoat()
