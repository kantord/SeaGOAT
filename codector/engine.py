"""
    This module allows you to use Codector as a library
"""

from pathlib import Path
import hashlib
from typing import Dict, List, Set
from typing_extensions import TypedDict

from tqdm import tqdm
from gitdb.db.loose import os
import appdirs
import chromadb
from chromadb.errors import IDAlreadyExistsError
from codector.cache import Cache

from codector.repository import Repository
from codector.result import Result
from codector.file import File

CACHE_FORMAT_VERSION = 14


RepositoryData = TypedDict(
    "RepositoryData",
    {
        "last_analyzed_version_of_branch": Dict[str, str],
        "required_commits": Set[str],
        "commits_already_analyzed": Set[str],
        "file_data": Dict[str, File],
        "sorted_files": List[str],
        "chunks_already_analyzed": Set[str],
    },
)


class Engine:
    """
    A search engine for a code repository
    """

    def __init__(self, path: str):
        """
        Initializes the library
        """
        self.path = path
        self.query_string = ""
        self._results = []

        self._chroma_client = chromadb.Client(
            chromadb.Settings(
                chroma_db_impl="duckdb+parquet",
                persist_directory=str(self._get_cache_folder()),
            )
        )
        self._chroma_collection = self._chroma_client.get_or_create_collection(
            name="code_data"
        )
        self._cache = Cache[RepositoryData](
            self._get_cache_folder() / "cache",
            {
                "last_analyzed_version_of_branch": {},
                "required_commits": set(),
                "commits_already_analyzed": set(),
                "file_data": {},
                "sorted_files": [],
                "chunks_already_analyzed": set(),
            },
        )
        self._cache.load()
        self.repository = Repository(path, self._cache)

    def _get_cache_folder(self):
        cache_folder = self._get_cache_root() / self._get_project_hash()
        cache_folder.mkdir(parents=True, exist_ok=True)

        return cache_folder

    def _get_cache_root(self):
        return Path(
            appdirs.user_cache_dir(
                "codector-pytest" if "PYTEST_CURRENT_TEST" in os.environ else "codector"
            )
        )

    def _get_project_hash(self):
        normalized_path = Path(self.path).expanduser().resolve()
        text = f"""
        Cache version: {CACHE_FORMAT_VERSION}
        Normalized path: {normalized_path}
        """

        return hashlib.sha256(text.encode()).hexdigest()

    def analyze_codebase(self):
        self.repository.analyze_files()
        self._create_vector_embeddings()

    def _add_to_collection(self, chunk):
        try:
            self._chroma_collection.add(
                ids=[chunk.chunk_id],
                documents=[chunk.chunk],
                metadatas=[{"path": chunk.path, "line": chunk.codeline}],
            )
        except IDAlreadyExistsError:
            pass

    def _create_vector_embeddings(self):
        chunks_to_process = []
        for file in self.repository.top_files()[:40]:
            for chunk in file.get_chunks():
                chunks_to_process.append(chunk)

        for chunk in tqdm(chunks_to_process, desc="Analyzing source code"):
            if chunk.chunk_id in self._cache.data["chunks_already_analyzed"]:
                continue

            self._add_to_collection(chunk)
            self._cache.data["chunks_already_analyzed"].add(chunk.chunk_id)

        self._cache.persist()
        self._chroma_client.persist()

    def query(self, query: str):
        self.query_string = query

    def fetch(self):
        chromadb_results = [
            self._chroma_collection.query(query_texts=[self.query_string], n_results=15)
        ]
        self._results = (
            chromadb_results[0]["metadatas"][0]
            if chromadb_results[0]["metadatas"]
            else None
        ) or []

    def get_results(self):
        path_order = []
        formatted_results = {}

        for metadata in self._results:
            path = str(metadata["path"])
            line = int(metadata["line"])
            if path not in path_order:
                path_order.append(path)

            if path not in formatted_results:
                formatted_results[path] = Result(path, Path(self.path) / path)
            formatted_results[path].add_line(line)

        return [formatted_results[path] for path in path_order]
