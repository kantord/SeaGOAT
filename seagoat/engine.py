"""
    This module allows you to use seagoat as a library
"""
import asyncio
from concurrent.futures import ThreadPoolExecutor
from functools import partial
from itertools import chain
from pathlib import Path
from typing import Dict
from typing import List
from typing import Set

import nest_asyncio
from tqdm import tqdm
from typing_extensions import TypedDict

from seagoat.cache import Cache
from seagoat.file import File
from seagoat.repository import Repository
from seagoat.sources import chroma
from seagoat.sources import ripgrep


RepositoryData = TypedDict(
    "RepositoryData",
    {
        "last_analyzed_version_of_branch": Dict[str, str],
        "required_commits": Set[str],
        "commits_already_analyzed": Set[str],
        "file_data": Dict[str, File],
        "sorted_files": List[str],
        "chunks_already_analyzed": Set[str],
    },
)


nest_asyncio.apply()


class Engine:
    """
    A search engine for a code repository
    """

    def __init__(self, path: str):
        """
        Initializes the library
        """
        self.path = path
        self.query_string = ""
        self._results_from_chromadb = []
        self._results = []
        self._cache = Cache[RepositoryData](
            "cache",
            Path(path),
            {
                "last_analyzed_version_of_branch": {},
                "required_commits": set(),
                "commits_already_analyzed": set(),
                "file_data": {},
                "sorted_files": [],
                "chunks_already_analyzed": set(),
            },
        )
        self._cache.load()
        self.repository = Repository(path, self._cache)
        self._fetchers = {
            "async": [
                ripgrep.initialize(self.repository),
            ],
            "sync": [
                chroma.initialize(self.repository),
            ],
        }

    def analyze_codebase(self):
        self.repository.analyze_files()
        self._create_vector_embeddings()

    def _add_to_collection(self, chunk):
        for source in chain(*self._fetchers.values()):
            source["cache_chunk"](chunk)

    def _create_vector_embeddings(self):
        chunks_to_process = []
        minimum_files_to_analyze = max(40, int(len(self.repository.top_files()) * 0.2))
        for file in self.repository.top_files()[:minimum_files_to_analyze]:
            for chunk in file.get_chunks():
                chunks_to_process.append(chunk)
        for chunk in tqdm(chunks_to_process, desc="Analyzing source code"):
            if chunk.chunk_id in self._cache.data["chunks_already_analyzed"]:
                continue

            self._add_to_collection(chunk)
            self._cache.data["chunks_already_analyzed"].add(chunk.chunk_id)

        self._cache.persist()

        for source in chain(*self._fetchers.values()):
            source["persist"]()

    def query(self, query: str):
        self.query_string = query

    async def fetch(self):
        self._results = []
        executor = ThreadPoolExecutor(max_workers=1)
        loop = asyncio.get_event_loop()
        async_tasks = [
            loop.run_in_executor(executor, partial(source["fetch"], self.query_string))
            for source in self._fetchers["async"]
        ]

        for source in self._fetchers["sync"]:
            self._results.extend(source["fetch"](self.query_string))

        results = await asyncio.gather(*async_tasks)
        for result in results:
            self._results.extend(result)

    def fetch_sync(self):
        loop = asyncio.get_event_loop()
        loop.run_until_complete(self.fetch())

    def get_results(self):
        merged_results = {}

        for result_item in self._results:
            if result_item.path not in merged_results:
                merged_results[result_item.path] = result_item
                continue

            merged_results[result_item.path].extend(result_item)

        sorted_by_score = sorted(
            merged_results.values(),
            key=lambda x: x.get_best_score(self.query_string),
        )
        position_by_result = {item: i for i, item in enumerate(sorted_by_score)}
        position_by_file = {
            file.path: i for i, file in enumerate(self.repository.top_files())
        }

        return list(
            sorted(
                merged_results.values(),
                key=lambda x: (
                    0.8 * position_by_result[x] + 0.2 * position_by_file[x.path]
                ),
            )
        )
