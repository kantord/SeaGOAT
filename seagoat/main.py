from functools import cache
import os
import sys
import click
from pygments import highlight
from pygments.lexers import get_lexer_for_filename
from pygments.formatters import TerminalFormatter
from pygments.lexers.javascript import JavascriptLexer, TypeScriptLexer

from seagoat.engine import Engine


@cache
def get_highlighted_lines(file_name: str):
    with open(file_name, "r", encoding="utf-8") as source_code_file:
        code = source_code_file.read()

    if file_name.endswith(".md"):
        return code.splitlines()

    if file_name.endswith(".jsx"):
        lexer = JavascriptLexer()
    elif file_name.endswith(".tsx"):
        lexer = TypeScriptLexer()
    else:
        lexer = get_lexer_for_filename(file_name)

    result = highlight(code, lexer, TerminalFormatter())

    return result.splitlines()


def print_result_line(result, line, color_enabled):
    if color_enabled:
        highlighted_lines = get_highlighted_lines(str(result.full_path))
        print(
            f"{result.path}:{click.style(str(line), bold=True)}:{highlighted_lines[line - 1]}"
        )
    else:
        print(f"{result.path}:{line}:{result.line_texts[line - 1]}")


@click.command()
@click.argument("query")
@click.argument("repo_path", required=False, default=os.getcwd())
@click.option("--no-color", is_flag=True)
def seagoat(query, repo_path, no_color):
    """Query your codebase using vector embeddings"""
    my_seagoat = Engine(repo_path)
    my_seagoat.analyze_codebase()

    my_seagoat.query(query)
    my_seagoat.fetch_sync()
    results = my_seagoat.get_results()

    color_enabled = sys.stdout.isatty() and not no_color
    for result in results:
        for line in result.get_lines(query):
            print_result_line(result, line, color_enabled)


if __name__ == "__main__":
    # pylint: disable-next=no-value-for-parameter
    seagoat()
