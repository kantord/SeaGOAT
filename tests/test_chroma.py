import copy
from pathlib import Path
from unittest.mock import patch

import pytest

from seagoat.engine import Engine


def normalize_full_paths(data, repo):
    real_repo_path = repo.working_dir
    fake_repo_path = "/path/to/repo"
    deep_copy_of_data = copy.deepcopy(data)
    deep_copy_of_data["fullPath"] = deep_copy_of_data["fullPath"].replace(
        real_repo_path, fake_repo_path
    )

    return deep_copy_of_data


@pytest.fixture(autouse=True)
def use_real_db(real_chromadb):
    pass


@pytest.mark.asyncio
async def test_gets_data_using_vector_embeddings(repo, snapshot):
    seagoat = Engine(repo.working_dir)
    seagoat.analyze_codebase()
    my_query = "lightweight markup language"
    await seagoat.fetch(my_query)

    # Tests that results are sorted according to relevance
    assert seagoat.get_results()[0].path == "file1.md"

    # Tests that results are grouped by file
    assert len(set(result.path for result in seagoat.get_results())) == len(
        list(seagoat.get_results())
    )

    # Tests that file lines are included for each result
    assert [
        normalize_full_paths(result.to_json(my_query), repo)
        for result in seagoat.get_results()
    ] == snapshot


def test_allows_fetching_data_synchronously(repo):
    repo.add_file_change_commit(
        file_name="articles.txt",
        contents="Ã©kezetes, Italian food recipes, spaghetti, pomodoro, pepperoni\n",
        author=repo.actors["John Doe"],
        commit_message="Add italian food recipes",
        encoding="windows-1250",
    )
    repo.add_file_change_commit(
        file_name="vehicles.txt",
        contents="Ford",
        author=repo.actors["John Doe"],
        commit_message="Add vehicle information",
    )
    repo.add_file_change_commit(
        file_name="vehicles.txt",
        contents="motorbike, ford, mercedes\n",
        author=repo.actors["John Doe"],
        commit_message="Add vehicle information",
    )
    seagoat = Engine(repo.working_dir)
    seagoat.analyze_codebase()
    my_query = "tomato pizza"
    seagoat.fetch_sync(my_query, limit_clue=33)

    assert seagoat.get_results()[0].path == "articles.txt"


@pytest.mark.asyncio
async def test_considers_filename_in_results(repo):
    repo.add_file_change_commit(
        file_name="cooking_recipes.txt",
        contents="motorbike, ford, mercedes\n",
        author=repo.actors["John Doe"],
        commit_message=".",
    )
    repo.add_file_change_commit(
        file_name="vehicles.txt",
        contents="Ford",
        author=repo.actors["John Doe"],
        commit_message="Add vehicle information",
    )
    repo.add_file_change_commit(
        file_name="vehicles.txt",
        contents="motorbike, ford, mercedes\n",
        author=repo.actors["John Doe"],
        commit_message="Add vehicle information",
    )
    seagoat = Engine(repo.working_dir)
    seagoat.analyze_codebase()
    my_query = "dish_recipe.txt"
    await seagoat.fetch(my_query)

    assert seagoat.get_results()[0].path == "cooking_recipes.txt"


@pytest.mark.asyncio
async def test_considers_commit_messages(repo):
    repo.add_file_change_commit(
        file_name="vehicles_1.txt",
        contents="the the the",
        author=repo.actors["John Doe"],
        commit_message="pizza tomato salami recipe",
    )
    repo.add_file_change_commit(
        file_name="vehicles_2.txt",
        contents=".",
        author=repo.actors["John Doe"],
        commit_message="Add vehicle information",
    )
    repo.add_file_change_commit(
        file_name="vehicles_2.txt",
        contents="",
        author=repo.actors["John Doe"],
        commit_message="Add vehicle information",
    )
    seagoat = Engine(repo.working_dir)
    seagoat.analyze_codebase()
    my_query = "italian pomodoro pie with slices of cured meat"
    await seagoat.fetch(my_query)

    assert seagoat.get_results()[0].path == "vehicles_1.txt"


@pytest.mark.asyncio
async def test_truncates_very_long_lines(repo):
    repo.add_file_change_commit(
        file_name="articles.txt",
        contents=f"car {'the a about ' * 40} pizza recipe tomato italian pie\n",
        author=repo.actors["John Doe"],
        commit_message="Add italian food recipes",
    )
    repo.add_file_change_commit(
        file_name="vehicles.txt",
        contents="Ford tomato",
        author=repo.actors["John Doe"],
        commit_message="Add vehicle information",
    )
    repo.add_file_change_commit(
        file_name="vehicles.txt",
        contents="motorbike, ford, mercedes with tomato and cheese\n",
        author=repo.actors["John Doe"],
        commit_message="Add vehicle information",
    )
    seagoat = Engine(repo.working_dir)
    seagoat.analyze_codebase()
    my_query = "tomato pizza"
    await seagoat.fetch(my_query)

    assert seagoat.get_results()[0].path == "vehicles.txt"


@pytest.mark.asyncio
async def test_includes_all_matching_lines_from_line(repo):
    repo.add_file_change_commit(
        file_name="devices.txt",
        contents="""1: Nothing
        2: Google Pixel 2 Android
        3:
        4: Mango juice
        5: Fried potatoes
        6: Chicken wings
        7: Apple iPhone 12
        8: Pizza slices with pepperoni
        9: Samsung Galaxy S10
        10:
        11:
        12:
        13:
        14:
        """,
        author=repo.actors["John Doe"],
        commit_message="Add italian food recipes",
    )
    seagoat = Engine(repo.working_dir)
    seagoat.analyze_codebase()
    my_query = "smartphone"
    await seagoat.fetch(my_query)

    assert seagoat.get_results()[0].path == "devices.txt"
    assert set(seagoat.get_results()[0].get_lines(my_query)) == {
        1,
        2,
        4,
        6,
        7,
        8,
        9,
    }


@pytest.mark.asyncio
async def test_exact_matches_have_higher_score(repo):
    repo.add_file_change_commit(
        file_name="devices.txt",
        contents="""1: Nothing
        2: Google Pixel 2 Android
        3:
        4: Mango juice
        5: Fried potatoes
        6: Chicken wings
        7: Apple iPhone 12
        8: Pizza slices with pepperoni
        9: Samsung Galaxy S10
        10:
        11:

        13:
        14:
        """,
        author=repo.actors["John Doe"],
        commit_message="Add italian food recipes",
    )
    seagoat = Engine(repo.working_dir)
    seagoat.analyze_codebase()
    my_query = "apple      iphone 12"
    await seagoat.fetch(my_query)

    assert seagoat.get_results()[0].path == "devices.txt"
    assert set(seagoat.get_results()[0].get_lines(my_query)) == {7, 8, 12}


@pytest.mark.asyncio
async def test_chunks_are_persisted_between_runs(repo):
    repo.add_file_change_commit(
        file_name="articles.txt",
        contents="Italian food recipes, spaghetti, pomodoro, pepperoni\n",
        author=repo.actors["John Doe"],
        commit_message="Add italian food recipes",
    )
    repo.add_file_change_commit(
        file_name="vehicles.txt",
        contents="Ford",
        author=repo.actors["John Doe"],
        commit_message="Add vehicle information",
    )
    repo.add_file_change_commit(
        file_name="vehicles.txt",
        contents="motorbike, ford, mercedes\n",
        author=repo.actors["John Doe"],
        commit_message="Add vehicle information",
    )
    seagoat1 = Engine(repo.working_dir)
    with patch.object(
        seagoat1, "_add_to_collection", wraps=seagoat1._add_to_collection
    ) as mock_add_to_collection:
        seagoat1.analyze_codebase()
        await seagoat1.fetch("pomodoro spaghetti")
        assert mock_add_to_collection.call_count > 2
        assert seagoat1.get_results()[0].path == "articles.txt"
        del seagoat1

    seagoat2 = Engine(repo.working_dir)
    with patch.object(
        seagoat2, "_add_to_collection", wraps=seagoat2._add_to_collection
    ) as mock_add_to_collection:
        seagoat2.analyze_codebase()
        await seagoat2.fetch("pomodoro spaghetti")
        assert mock_add_to_collection.call_count == 0
        assert seagoat2.get_results()[0].path == "articles.txt"


@pytest.mark.asyncio
async def test_respects_limit_in_chromadb(repo):
    for i in range(100):
        repo.add_file_change_commit(
            file_name="devices2.txt",
            contents="food from an orchard, but not a banana or pear. It's red",
            author=repo.actors["John Doe"],
            commit_message=f"commit {i}",
        )
        repo.tick_fake_date(days=1)
    repo.tick_fake_date(days=1)
    repo.add_file_change_commit(
        file_name="devices.txt",
        contents="""banana, fruit, red, Macintosh, iPhone
        """
        * 6,
        author=repo.actors["John Doe"],
        commit_message="Add italian food recipes",
    )
    seagoat = Engine(repo.working_dir)
    seagoat.analyze_codebase()
    my_query = "apple"
    await seagoat.fetch(my_query, limit_clue=5)

    expected_files = {"devices.txt", "devices2.txt"}
    results_files = set(result.path for result in seagoat.get_results())

    assert expected_files.issubset(results_files)
    result_for_devices_txt = next(
        (result for result in seagoat.get_results() if result.path == "devices.txt"),
        None,
    )
    if result_for_devices_txt:
        assert len(result_for_devices_txt.get_lines(my_query)) == 6
    else:
        raise AssertionError("File 'devices.txt' not found in results.")


@pytest.mark.asyncio
async def test_does_not_crash_repo_when_files_are_deleted(repo):
    repo.add_file_change_commit(
        file_name="cooking_recipes.txt",
        contents="motorbike, ford, mercedes\n",
        author=repo.actors["John Doe"],
        commit_message=".",
    )
    seagoat = Engine(repo.working_dir)
    seagoat.analyze_codebase()
    my_query = "dish_recipe.txt"
    await seagoat.fetch(my_query)
    (Path(repo.working_dir) / "cooking_recipes.txt").unlink()

    seagoat.analyze_codebase()
    await seagoat.fetch(my_query)


@pytest.mark.asyncio
async def test_custom_ignore_patterns(repo, create_config_file):
    create_config_file(
        {
            "server": {"ignorePatterns": ["**/devices2.txt"]},
        }
    )
    repo.add_file_change_commit(
        file_name="foo/devices2.txt",
        contents="food from an orchard, but not a banana or pear. It's red",
        author=repo.actors["John Doe"],
        commit_message="commit",
    )
    repo.add_file_change_commit(
        file_name="devices.txt",
        contents="""banana, fruit, red, Macintosh, iPhone
        """
        * 6,
        author=repo.actors["John Doe"],
        commit_message="Add italian food recipes",
    )
    seagoat = Engine(repo.working_dir)
    seagoat.analyze_codebase()
    my_query = "apple"
    await seagoat.fetch(my_query, limit_clue=5)

    results_files = set(result.path for result in seagoat.get_results())

    assert "foo/devices2.txt" not in results_files


@pytest.mark.asyncio
async def test_user_defined_execution_provider(repo, create_config_file):
    create_config_file(
        {
            "server": {
                "chroma": {
                    "embeddingFunction": {
                        "name": "ONNXMiniLM_L6_V2",
                        "arguments": {"preferred_providers": ["CPUExecutionProvider"]},
                    },
                }
            }
        }
    )
    seagoat = Engine(repo.working_dir)
    seagoat.analyze_codebase()
    my_query = "apple"
    seagoat.fetch_sync(my_query)
