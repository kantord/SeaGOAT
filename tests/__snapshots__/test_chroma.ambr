# serializer version: 1
# name: test_gets_data_using_vector_embeddings
  list([
    dict({
      'blocks': list([
        dict({
          'lineTypeCount': dict({
            'result': 1,
          }),
          'lines': list([
            dict({
              'line': 3,
              'lineText': "HTML is for markup, but it's complicated. Markdown is simpler.",
              'resultTypes': list([
                'result',
              ]),
            }),
          ]),
        }),
      ]),
      'fullPath': '/path/to/repo/file1.md',
      'path': 'file1.md',
    }),
  ])
# ---
